src/spec.ts

```diff
@@ -120,18 +129,91 @@ export function normalize(spec: ExtendedSpec): Spec {
           hasRow ? { row: spec.encoding.row } : {},
           hasColumn ? { column: spec.encoding.column } : {}
         ),
-        spec: {
+        spec: normalizeUnitSpec({
           mark: spec.mark,
           encoding: encoding
-        }
+        })
       },
       spec.config ? { config: spec.config } : {}
     );
+}
+
+export function normalizeUnitSpec(spec: UnitSpec): Spec {
+  if (contains([ERRORBAR], spec.mark)) {
+    return normalizeCompositeUnitSpec(spec);
+  }
+  if (isRanged(spec.encoding)) {
+    return normalizeRangedUnitSpec(spec);
   }
+  return spec;
+}
 
+export function normalizeRangedUnitSpec(spec: UnitSpec): Spec {
+  if (spec.encoding) {
+    const hasX = has(spec.encoding, X);
+    const hasY = has(spec.encoding, Y);
+    const hasX2 = has(spec.encoding, X2);
+    const hasY2 = has(spec.encoding, Y2);
+    if ((hasX2 && !hasX) || (hasY2 && !hasY)) {
+      let normalizedSpec = duplicate(spec);
+      if (hasX2 && !hasX) {
+        normalizedSpec.encoding.x = duplicate(normalizedSpec.encoding.x2);
+        delete normalizedSpec.encoding.x2;
+      }
+      if (hasY2 && !hasY) {
+        normalizedSpec.encoding.y = duplicate(normalizedSpec.encoding.y2);
+        delete normalizedSpec.encoding.y2;
+      }
+
+      return normalizedSpec;
+    }
+  }
   return spec;
 }
 
+export function normalizeCompositeUnitSpec(spec: UnitSpec): Spec {
+  let layerSpec = extend(spec.name ? {name: spec.name} : {},
+    spec.description ? {description: spec.description} : {},
+    spec.data ? {data: spec.data} : {},
+    spec.transform ? {transform: spec.transform} : {},
+    spec.config ? {config: spec.config} : {}, {layers: []}
+  );
+  if (!spec.encoding) {
+    return layerSpec;
+  }
+  if (spec.mark === ERRORBAR) {
+    const ruleSpec = {
+      mark: RULE,
+      encoding: extend(
```

Comment:
If removing this chunk of code from thIs PR is troublesome, I'm fine with letting it be merged as is, as long as you send a follow up PR.  

If you want to do that, create another issue and assign to yourself, so that we can track and I'll ignore this part for now :)